### PR TITLE
Wire up cycle with data training and process

### PR DESCRIPTION
**Pull request by mccorby**

### CODE REVIEW

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt [hunk 1]
```diff
@@ -1,43 +1,83 @@
...
+        model: SyftModel,
+        plans: ConcurrentHashMap<String, Plan>,
+        clientConfig: ClientConfig
+    ) {
+        val destinationDir = localConfiguration.plansLocation
+
+        val plan = plans.toList().first().second
+        val loadData = mnistDataRepository.loadData()
+        repeat(EPOCHS) {
+            val output = plan.execute(model, loadData, clientConfig)
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:78]: You would need to call execute function on each batch/data point

Something like: 

ZippedData = zip(loadData.first,loadData.second)
zippedData.foreach{
plan.execute(model,it,clientConfig)
}

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt [hunk 2]
```diff
@@ -1,43 +1,79 @@
...
+    }
+
+    private fun trainingProcess(
+        model: SyftModel,
+        plans: ConcurrentHashMap<String, Plan>,
+        clientConfig: ClientConfig
+    ) {
+        val destinationDir = "/data/data/org.openmined.syft.demo/files/plans"
+
+        val plan = plans.toList().first().second
```
[the remark this answers is not in the document]
Comment (mccorby) [reply demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:75]: I'm having an issue with `softmax`. Not sure why.

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt [hunk 3]
```diff
@@ -1,43 +1,86 @@
...
+        plans: ConcurrentHashMap<String, Plan>,
+        clientConfig: ClientConfig
+    ) {
+        val destinationDir = localConfiguration.plansLocation
+
+        val plan = plans.toList().first().second
+        val loadData = mnistDataRepository.loadData()
+        repeat(EPOCHS) {
+            val zipData = loadData.first zip loadData.second
+            zipData.forEach {
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:79]: Val zipData declaration shouldn't be inside its the same everytime

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (vkkhare) [reply demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:75]: Why do you need to evaluate softmax? It's already done in the plan. We directly get the accuracy values as an output.

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt [hunk 4]
```diff
@@ -71,17 +75,31 @@ class FederatedCycleViewModel(
...
+                        clientConfig.copy(batchSize = 1)
+                    )?.toTuple()
+                    output?.let { outputResult ->
+                        val paramSize = model.modelState!!.syftTensors.size
+                        val beginIndex = output.size - paramSize
+                        val updatedParams =
+                                outputResult.slice(beginIndex until beginIndex + paramSize - 1)
+                        model.updateModel(updatedParams.map { it.toTensor() })
+                        val result =
+                                (output[0] as IValue).toTensor().dataAsFloatArray.last().toString()
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:97]: We don't need as Ivalue anymore. The will only return a single Ivalue array

File: demo-app/src/main/java/org/openmined/syft/demo/ui/MainActivity.kt [hunk 5]
```diff
@@ -1,53 +1,75 @@
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         val binding: ActivityMainBinding =
                 DataBindingUtil.setContentView(this, R.layout.activity_main)
         setSupportActionBar(toolbar)
-        binding.viewModel = initiateViewModel("192.168.1.13:5000")
+        binding.lifecycleOwner = this
+        binding.viewModel = initiateViewModel("10.0.2.2:5000")
+
+        (binding.viewModel as FederatedCycleViewModel).processState.observe(this, Observer { onProcessStateChanged(it) })
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/MainActivity.kt:35]: Maybe break the line here?

File: demo-app/src/main/java/org/openmined/syft/demo/ui/MainActivity.kt [hunk 6]
```diff
@@ -1,53 +1,75 @@
...
 class MainActivity : AppCompatActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         val binding: ActivityMainBinding =
                 DataBindingUtil.setContentView(this, R.layout.activity_main)
         setSupportActionBar(toolbar)
-        binding.viewModel = initiateViewModel("192.168.1.13:5000")
+        binding.lifecycleOwner = this
+        binding.viewModel = initiateViewModel("10.0.2.2:5000")
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/MainActivity.kt:33]: Also can you add a line chart from https://github.com/PhilJay/MPAndroidChart? It is pretty easy and just 3-4 lines. Instead of showing the logger we can have dynamically populated graph

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt [hunk 7]
```diff
@@ -1,43 +1,116 @@
...
+                        trainingBatch,
+                        clientConfig.copy(batchSize = 1)
+                    )?.toTuple()
+                    output?.let { outputResult ->
+                        val paramSize = model.modelState!!.syftTensors.size
+                        val beginIndex = output.size - paramSize
+                        val updatedParams =
+                                outputResult.slice(beginIndex until beginIndex + paramSize - 1)
+                        model.updateModel(updatedParams.map { it.toTensor() })
+                        result.add(outputResult[0].toTensor().dataAsFloatArray.last().toString())
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:95]: We dont need to convert to string here if we are sending floats to graph

File: demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt [hunk 8]
```diff
@@ -1,43 +1,116 @@
...
+            val zipData = loadData.first zip loadData.second
+            repeat(EPOCHS) {
+                postLog("Starting epoch ${it + 1}")
+                zipData.forEach { trainingBatch ->
+                    val output = plan.execute(
+                        model,
+                        trainingBatch,
+                        clientConfig.copy(batchSize = 1)
+                    )?.toTuple()
+                    output?.let { outputResult ->
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/ui/FederatedCycleViewModel.kt:89]: I ran the demo and everything looks great but can you populate the graph dynamically after say every 10 epochs rather than waiting till the end for training to finish. I would like to see if the loss error actually reduces over time. Also is the training data the complete MNIST data?
